src/coreclr/jit/codegenarm64.cpp

```diff
@@ -4412,57 +4412,166 @@ void CodeGen::genCodeForCompare(GenTreeOp* tree)
+    var_types op2Type   = genActualType(op2->TypeGet());
+    emitAttr  cmpSize   = EA_ATTR(genTypeSize(op1Type));
     regNumber targetReg = tree->GetRegNum();
     regNumber srcReg1   = genConsumeReg(op1);
 
-    if (tree->OperIs(GT_SELECT))
+    // No float support or swapping op1 and op2 to generate cmp reg, imm.
+    assert(!varTypeIsFloating(op2Type));
+    assert(!op1->isContainedIntOrIImmed());
+
+    // For the ccmp flags, invert the condition of the compare.
+    insCflags cflags = InsCflagsForCcmp(InsCondForCompareOp(tree));
+
+    if (op2->isContainedIntOrIImmed())
     {
-        regNumber srcReg2 = genConsumeReg(op2);
-        emit->emitIns_R_R_R_COND(INS_csel, cmpSize, targetReg, srcReg1, srcReg2, cond);
-        regSet.verifyRegUsed(targetReg);
+        GenTreeIntConCommon* intConst = op2->AsIntConCommon();
+        emit->emitIns_R_I_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, (int)intConst->IconValue(), cflags, cond);
     }
     else
     {
-        assert(!varTypeIsFloating(op2Type));
-        // We don't support swapping op1 and op2 to generate cmp reg, imm.
-        assert(!op1->isContainedIntOrIImmed());
-        // This should not be generating into a register.
-        assert(targetReg == REG_NA);
+        regNumber srcReg2 = genConsumeReg(op2);
+        emit->emitIns_R_R_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, srcReg2, cflags, cond);
+    }
+
+    // Are we evaluating this into a register?
+    if (targetReg != REG_NA)
+    {
+        inst_SETCC(GenCondition::FromRelop(tree), tree->TypeGet(), targetReg);
+        genProduceReg(tree);
+    }
+}
 
-        // For the ccmp flags, get the condition of the compare.
-        insCflags cflags = InsCflagsForCcmp(InsCondForCompareOp(tree));
+//------------------------------------------------------------------------
+// genCodeForContainedCompareChain: Produce code for a chain of conditional compares.
+//
+// Only generates for contained nodes. Nodes that are not contained are assumed to be
+// generated as part of standard tree generation.
+//
+// Arguments:
+//    tree - the node. Either a compare or a tree of compares connected by ANDs.
+//    inchain - whether a contained chain is in progress.
+//    prev - If a chain is in progress, the condition of the previous compare.
+// Return:
+//    The last compare node generated.
+//
+void CodeGen::genCodeForContainedCompareChain(GenTreeOp* tree, bool* inchain, insCond* prevcond)
+{
+    if (tree->OperIs(GT_AND))
+    {
+        GenTreeOp* op1 = tree->gtGetOp1()->AsOp();
+        GenTreeOp* op2 = tree->gtGetOp2()->AsOp();
 
-        if (op2->isContainedIntOrIImmed())
+        if (!tree->isContained())
         {
-            GenTreeIntConCommon* intConst = op2->AsIntConCommon();
-            emit->emitIns_R_I_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, (int)intConst->IconValue(), cflags, cond);
+            // An And that is not contained should not have any contained children.
+            assert(!op1->isContained() && !op2->isContained());
```

Comment:
This code should not make any assumptions about this.

Reply:
The problem is handling that case.

Firstly that would mean having extra logic in the AND code generation to handle contained children. Do we then assume the contained children of the AND can be of any node type?

Then should I apply compare chain logic to it? If op1 is a compare chain, then op2 should probably generate a conditional compare, and then the AND needs to generate into a register. Alternatively, just do the easy way of generate op1 into a register, op2 into a register and do a normal AND.

The problem is, those cases aren't going to be happen (due to the lower phase not creating non-contained ANDs that have contained children).

So maybe the answer is:
in the code generation for and, if it has contained children, then generate those children into registers and then plant the AND as normal

Reply:
> No; this should be left up to codegen of AND as it exists today.

There will still need to be some changes in the codegen for `AND`....

in `CodeGen::genCodeForBinary()` for Arm64 it only handles contained for MUL: 
`if (op2->OperIs(GT_MUL) && op2->isContained())`

> They should be opaque to the grandparent node

Agreed (and the subsequent comments too)

Reply:
> There will still need to be some changes in the codegen for AND....

I agree this will be necessary if we are going to add support for `AND` to consume compares via `ccmp` instead of by materializing the truth value into a register. However, if we are leaving the support for `SELECT` only then I don't see why it should be necessary, but maybe I am missing something.

I guess the way I would shape this is the following:
* Make `genCodeForConditionalCompare` always produce a value into the flags, never into a register. Assert that it is only called on contained `AND` and compare nodes of the right shapes. In the base cases, we still need to call `genConsumeReg` on operands and get values from registers.
* Change `genCodeForSelect` to satisfy the above: if the cond op is contained, then call `genCodeForConditionalCompare`, and otherwise call `genConsumeReg` and generate code to compare the register with 0
* (Optional) Make a similar change in codegen for `GT_AND` to support contained comparisons there via `ccmp` as well. This will need to materialize a truth value.

The way the current code is shaped might also be fine, but you are probably missing some handling for the base cases (where the operands are no longer contained).

Reply:
That logic doesn't quite work. For a 'SELECT' node, if the conditional is contained then it could be a 'CMP' node or an 'AND' node. So we still need something to handle both.

I'll refine the existing code using some of the above, and add some code in the AND and then see where that gets us....

Reply:
My suggestion was to handle the contained AND in `genCodeForConditionalCompare`, in the same way you are doing now. 
It is fine to assert that you _only_ see contained ANDs here.

Reply:
This part is updated.

Reply:
It looks great now, but I'll have to spend a bit of time on this to make sure that this handles interference checking with `IsSafeToContainMem` correctly.